### PR TITLE
[12.0][FIX] - report_py3o: run libreoffice in an isolated user installation

### PR DESCRIPTION
**Pull request by sbejaoui**
Bug when more than one conversion print is launched within the same libreoffice instance.

The standard behavior of libreoffice when a user open it while another instance is
running is to show a new window and throw an error if a new instance is forced within
the same user installation [see](https://bugs.documentfoundation.org/show_bug.cgi?id=37531).

This implies a bug in report_py3o module when we call libreoffice at the same time for
different documents.

To reproduce this bug:

**Case 1:**
1.  Simultaneously print two documents.

**Case 2:**
1.  Run print jobs using job_queue module
2.  Manually print another document

**Case 3:**
2.  Open libreoffice
3.  Print a py3o report

This PR creates a temporary user installation for each libreoffice conversion to bypass this limitation.

cc/ @sbidoul

### CODE REVIEW

File: report_py3o/models/py3o_report.py [hunk 1]
```diff
@@ -265,18 +267,23 @@ def _convert_single_report(self, result_path, model_instance, data):
...
+        return (
+            [
+                lo_bin,
+                '-env:UserInstallation=file:%s' % tmp_user_installation,
+                '--headless',
+                '--convert-to',
+                self.ir_actions_report_id.py3o_filetype,
+                result_path,
+            ],
+            tmp_user_installation,
```
Comment (sbidoul) [inline report_py3o/models/py3o_report.py]: Why do we need to return tmp_user_installation to the caller? Can't we clean it here?

**Comment by sbidoul** [comment]
@hbrunn @lmignon do you have an opinion about this? Have you ever encountered this issue?

File: report_py3o/models/py3o_report.py (hunk 1 above)
Comment (hbrunn) [reply report_py3o/models/py3o_report.py]: no, we need this to exist for the actual call to libreoffice to succeed. But I'm also concerned about breaking the API here, I'd prefer to see the the temporary directory passed in an optional keyword argument. If passed, we add the parameter to the command, and the caller is responsible for cleaning up.

**Comment by hbrunn** [comment]
I'm a little bit surprised by the repeated mention of the fusion server, as this is exactly the code path where we don't use the fusion server.

Myself I never encountered the issue described, but then we also use this only for a handful of reports that are very unlikely to be run anywhere close to parallel. That said, the bug description makes sense to me, I immediately believe this can be a problem. So as soon as this doesn't break the API, I'd be fine with the change.

File: report_py3o/models/py3o_report.py (hunk 1 above)
Comment (sbidoul) [reply report_py3o/models/py3o_report.py]: right, I skimmed on it too quickly. So yes, better doing a `with TemporaryDirectory()` in the caller and pass the temp dir to `_convert_single_report`.

**Comment by sbejaoui** [comment]
> I'm a little bit surprised by the repeated mention of the fusion server, as this is exactly the code path where we don't use the fusion server.

Thank you for your review,

I misunderstood the usage of the fusion server as I am not familiar with the module. I updated the PR description.

File: report_py3o/models/py3o_report.py (hunk 1 above)
Comment (sbejaoui) [reply report_py3o/models/py3o_report.py]: I pushed a new version to not break the API

**Review by lmignon — APPROVED** [review]
Thank you for this fix @sbejaoui. All this makes sens according to https://ask.libreoffice.org/en/question/14841/how-can-i-use-soffice-from-the-command-line-when-quickstarter-is-running/

File: report_py3o/models/py3o_report.py [hunk 2]
```diff
@@ -243,40 +243,49 @@ def _create_single_report(self, model_instance, data):
...
+    ):
         """Return a command list suitable for use in subprocess.call"""
         lo_bin = self.ir_actions_report_id.lo_bin_path
         if not lo_bin:
             raise RuntimeError(
                 _("Libreoffice runtime not available. "
                   "Please contact your administrator.")
             )
-        return [
+        params = [
```
Comment (sbidoul) [inline report_py3o/models/py3o_report.py]: nit: I'd name this variable cmd to match the method name.

**Review by sbidoul — APPROVED** [review]

**Review by hbrunn — APPROVED** [review]

File: report_py3o/models/py3o_report.py (hunk 2 above)
Comment (hbrunn) [reply report_py3o/models/py3o_report.py]: if you do this nit, please also add a comma after `None`, but I don't insist on it. Who does the honors?

Comment (sbidoul) [reply report_py3o/models/py3o_report.py]: actually, black does not add a comma in this particular case

**Pull request merged**